Documentation/design-docs/variabletracking.md

```diff
@@ -0,0 +1,314 @@
+
+This is done in `CodeGen::genSetScopeInfoUsingsiScope()`
+
+Variable Live Range Structure
+--------
+
+### VariableLiveRange
+
+This class is used to represent an uninterruptible portion of variable live in one location.
+Owns two emitLocations indicating the first instructions that make it valid and invalid respectively.
+It also has the location of the variable in this range, which is stored at the moment of being created.
+
+To save space, we save the ending emitLocation only when the variable is being moved or its liveness change, which could happen many blocks after it was being born.
+This means a `VariableLiveRange` native offsets range could be longer than a `BasicBlock`.
+
+### VariableLiveDescriptor
+
+This class is used to represent the liveness of a variable.
+It has a list of `VariableLiveRange`s and common operations to update the variable liveness.
+
+### VariableLiveKeeper
+
+It holds an array of `VariableLiveDescriptor`s, one per variable that is being tracked.
+The index of each variable inside this array is the same as in `Compiler::lvaTable`.
+
+We could have `VariableLiveDescriptor` inside each LclVarDsc or an array of `VariableLiveDescriptor` inside `Compiler`, but the intention is to move code out of `Compiler` and not increse it.
+
+Generating Variable Live Range
+--------
+
+In order to use Variable Live Range for tracking variables location, the Flag `USING_VARIABLE_LIVE_RANGE` in codegeninterface.h should be defined.
+
+### For optimized code
+
+In `genCodeForBBList()`, we start generating code for each block dealing with some cases.
+
+On `BasicBlock` boundaries:
+
+-   `BasicBlock`s beginning:
+
+    -   If a variable has an open `VariableLiveRange` but the location is different than what is expected to be in this block we update it (close the existing one and create another).
+        This could happen because a variable could be in the `bbLiveOut` of a block and in the `BasicBlock`s `bbLiveOut` of the next one, but that doesn't mean that the execution thread would go from one inmeadiatly to the next one.
+        For this kind of cases another block that moves the variable from its original to the expected position is created.
+        This is handled in `LinearScan::recordVarLocationsAtStartOfBB(BasicBlock* bb)`.
+
+    -   If a variable has not an open `VariableLiveRange` and is in `bbLiveIn`, we open one.
+        This is done in `genUpdateLife` inmeadiatly after the the previous method is called.
+
+    -   If a variable has an open `VariableLiveRange` and is not in `bbLiveIn`, we close it.
+        This is handled considered in `genUpdateLife` too.
+
+-   last `BasicBlock`s ending:
+
+    -   We close every open `VariableLiveRange`.
+        This is handled in `genCodeForBBList` when iterating the blocks, after the code for each block is done.
+
+For each instruction in `BasicBlock`:
+-   a `VariableLiveRange` is opened for each variable that is being born and not is becoming dead at the same instruction
```

Comment:
"being born and not is becoming dead at the same instruction"

Can you explain a little more about this case? E.g., what's an example? Why not report this?